lib/client.js

```diff
@@ -1278,6 +1280,149 @@ function _resume (client, sync) {
+
+    if (socket.destroyed) {
+      return
+    }
+
+    if (bytesWritten === 0) {
+      if (expectsPayload) {
+        // https://tools.ietf.org/html/rfc7230#section-3.3.2
+        // A user agent SHOULD send a Content-Length in a request message when
+        // no Transfer-Encoding is sent and the request method defines a meaning
+        // for an enclosed payload body.
+
+        socket.write(`${header}content-length: 0\r\n\r\n\r\n`, 'ascii')
+      } else {
+        socket.write(`${header}\r\n`, 'ascii')
+      }
+    } else if (contentLength === null) {
+      socket.write('\r\n0\r\n\r\n', 'ascii')
+    }
+
+    if (socket[kParser].timeout && socket[kParser].timeoutType === TIMEOUT_HEADERS) {
+      // istanbul ignore else: only for jest
+      if (socket[kParser].timeout.refresh) {
+        socket[kParser].timeout.refresh()
+      }
+    }
+
+    resume(client)
+  }
+
+  let thrownError = null
+  try {
+    socket.on('error', onFinished)
+    const isSync = !!body[Symbol.iterator]
+    iterator = isSync ? body[Symbol.iterator]() : body[Symbol.asyncIterator]()
+    while (true) {
+      let nextResult
+      try {
+        nextResult = iterator.next()
+        if (!isSync) {
+          nextResult = await nextResult
+        }
+      } catch (err) {
+        // throw from next is completion
+        iteratorNormalCompletion = true
+        throw err
+      }
+
+      const { value: chunk, done } = nextResult
+      if (done || finished || socket.destroyed) {
+        if (done) {
+          // done is normal
+          iteratorNormalCompletion = true
+        }
+        return
+      }
+
+      const len = Buffer.byteLength(chunk)
+      if (!len) {
+        if (isSync) {
+          // await as otherwise this could run forever
+          await Promise.resolve()
+        }
+        continue
+      }
+
+      // TODO: What if not ended and bytesWritten === contentLength?
+      // We should defer writing chunks.
+      if (contentLength !== null && bytesWritten + len > contentLength) {
+        if (client[kStrictContentLength]) {
+          throw new RequestContentLengthMismatchError()
+        }
+
+        process.emitWarning(new RequestContentLengthMismatchError())
+      }
+
+      if (bytesWritten === 0) {
+        if (!expectsPayload) {
+          socket[kReset] = true
+        }
+
+        if (contentLength === null) {
+          socket.write(`${header}transfer-encoding: chunked\r\n`, 'ascii')
+        } else {
+          socket.write(`${header}content-length: ${contentLength}\r\n\r\n`, 'ascii')
+        }
+      }
+
+      if (contentLength === null) {
+        socket.write(`\r\n${len.toString(16)}\r\n`, 'ascii')
+      }
+
+      bytesWritten += len
+
+      if (socket.write(chunk)) {
+        continue
+      } else if (socket.destroyed) {
+        return
+      } else {
+        await EventEmitter.once(socket, 'drain')
+      }
+    }
+  } catch (err) {
+    thrownError = err
+  } finally {
+    onFinished(thrownError)
+  }
+}
```

Comment:
Implementation is simpler if you just use a for await loop instead of manually handling the iterator.

Reply:
Yes of course, but then I won't be able to close the iterator on socket error as I won't have access to the iterator.

Reply:
Technically, now that I think about it, I may have over-complicated it as once the next iteration beginds ends the for...await will just exit by itself anyway.

Reply:
that's what happens when you write code in the middle of the night...